core-bundle/src/Controller/BackendPreviewSwitchController.php

```diff
@@ -0,0 +1,205 @@
+ */
+
+namespace Contao\CoreBundle\Controller;
+
+use Contao\BackendUser;
+use Contao\CoreBundle\Framework\ContaoFramework;
+use Contao\CoreBundle\Security\Authentication\FrontendPreviewAuthenticator;
+use Contao\CoreBundle\Security\Authentication\Token\TokenChecker;
+use Contao\Date;
+use Doctrine\DBAL\Connection;
+use Doctrine\DBAL\FetchMode;
+use Symfony\Component\HttpFoundation\JsonResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Routing\Annotation\Route;
+use Symfony\Component\Routing\RouterInterface;
+use Symfony\Component\Security\Core\Security;
+use Symfony\Component\Security\Csrf\CsrfTokenManagerInterface;
+use Twig\Environment as TwigEnvironment;
+use Twig\Error\Error as TwigError;
+
+/**
+ * This controller serves for the back end preview toolbar by providing the following ajax endpoints:
+ * a) Return the toolbar html (dispatched in an ajax request to allow lazy loading and force back end scope)
+ * b) Provide members' usernames for the datalist
+ * c) Process the switch action (i.e. log in a specific front end user).
+ *
+ * @Route(defaults={"_scope" = "backend"})
+ */
+class BackendPreviewSwitchController
+{
+    private $contaoFramework;
+
+    private $frontendPreviewAuthenticator;
+
+    private $tokenChecker;
+
+    private $connection;
+
+    private $security;
+
+    private $twig;
+
+    private $tokenManager;
+
+    private $csrfTokenName;
+
+    private $router;
+
+    public function __construct(
+        ContaoFramework $contaoFramework,
+        FrontendPreviewAuthenticator $frontendPreviewAuthenticator,
+        TokenChecker $tokenChecker,
+        Connection $connection,
+        Security $security,
+        TwigEnvironment $twig,
+        RouterInterface $router,
+        CsrfTokenManagerInterface $tokenManager,
+        string $csrfTokenName
+    ) {
+        $this->contaoFramework = $contaoFramework;
+        $this->frontendPreviewAuthenticator = $frontendPreviewAuthenticator;
+        $this->tokenChecker = $tokenChecker;
+        $this->connection = $connection;
+        $this->security = $security;
+        $this->twig = $twig;
+        $this->router = $router;
+        $this->tokenManager = $tokenManager;
+        $this->csrfTokenName = $csrfTokenName;
+    }
+
+    /**
+     * @Route("/contao/preview_switch", name="contao_backend_preview_switch")
+     */
+    public function __invoke(Request $request): Response
+    {
+        $this->contaoFramework->initialize();
+
+        $user = $this->security->getUser();
+
+        if (!($user instanceof BackendUser) || !$request->isXmlHttpRequest()) {
+            return new Response('Bad Request', Response::HTTP_BAD_REQUEST);
+        }
+
+        if ($request->isMethod('GET')) {
+            try {
+                $toolbar = $this->renderToolbar($user);
+            } catch (TwigError $e) {
+                return new Response('', Response::HTTP_INTERNAL_SERVER_ERROR);
```

Comment:
I think I would force `renderToolbar()` to return a `string` and catch the error inside there and return a general `could not render twig template xy beacuse of: $e->getMessage()`. It shouldn't happen in production anyway as you're not adjusting templates there (especially not twig ones). So it's more of a message when adjusting the template in dev mode and I think in that case a proper error message would be helpful.